pydantic/_internal/_validate_call.py

```diff
@@ -97,3 +102,105 @@ def __call__(self, *args: Any, **kwargs: Any) -> Any:
         if self.__return_pydantic_validator__:
             return self.__return_pydantic_validator__(res)
         return res
+
+
+class ValidateCallInfo(TypedDict):
+    validate_return: bool
+    config: ConfigDict | None
+    function: Callable[..., Any]
+    local_namspace: dict[str, Any] | None
+
+
+def _is_wrapped_by_validate_call(obj: object) -> bool:
+    return hasattr(obj, '__pydantic_validate_call_info__')
+
+
+def collect_validate_call_info(namespace: dict[str, Any]):
+    validate_call_infos = {}
+    for name, func in namespace.items():
+        if _is_wrapped_by_validate_call(func):
+            validate_call_infos[name] = func.__pydantic_validate_call_info__
+
+    return validate_call_infos
+
+
+@lru_cache(maxsize=None)
+def _add_unique_postfix(name: str):
+    """Used to prevent namespace collision."""
+    from uuid import uuid4
+
+    postfix = str(uuid4()).replace('-', '_')
+    return f'{name}_{postfix}'
+
+
+def _replicate_validate_call(function: Callable[..., Any], info: ValidateCallInfo) -> Callable[..., Any]:
+    """When normally calling `validate_call`, we use the namespace of the frame that called it as local_ns.
+    This function mock that behavior by calling `validate_call` inside a new frame where we have copied all
+    local variables into.
+    """
+    namespace = info['local_namspace']
+
+    locals_name = _add_unique_postfix('locals')
+    parent_name = _add_unique_postfix('parent')
+    info_name = _add_unique_postfix('info')
+    function_name = _add_unique_postfix('function')
+    item_name = _add_unique_postfix('item')
+
+    from ..validate_call_decorator import validate_call
+
+    result_ns = {locals_name: namespace, 'validate_call': validate_call, info_name: info, function_name: function}
+
+    # Note: here the parent scope has to be class, not function, because since Python 3.13
+    #       the `locals()` cannot be mutate inside function frame.
+    # See: https://docs.python.org/3.13/library/functions.html#locals
+    #      https://docs.python.org/3.13/glossary.html#term-optimized-scope
+    exec(
+        f"""
+class {parent_name}:
+    for {item_name} in {locals_name}.items():
+        locals()[{item_name}[0]] = {item_name}[1]
+    del {item_name}
+
+    {function_name} = validate_call(config={info_name}['config'], validate_return={info_name}['validate_return'])({function_name})
+""",
+        result_ns,
+    )
```

Comment:
Not loving this; is there any way to avoid the exec here?

If we _need_ to keep this, I definitely feel like this needs more justification in a comment for why we are doing something with code generation like this and why it is necessary to do so. I see the comment above explaining that we are defining it in a new frame but it's less clear to me why it needs to be done via codegen and not via more "standard" python. Or what the purpose of redefining it like this is in the first place